### PR TITLE
Add the runtime for the miner

### PR DESCRIPTION
**Pull request by SWvheerden**
## Description
This PR adds in the miner runtime as well as the miner base-node event stream.

## Motivation and Context
This allows the miner to run on its own thread, as well as listen to the base node when it receives a new block. When a valid new block has been received it will restart the miner on the new tip.

## Types of changes
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

### CODE REVIEW

File: applications/tari_base_node/src/main.rs [hunk 1]
```diff
@@ -131,7 +131,13 @@ fn main() {
 
     // Configure the shutdown daemon to listen for CTRL-C
     let flag = node.get_flag();
-    handle_ctrl_c(&rt, flag);
+    handle_ctrl_c(&rt, flag.clone());
```
Comment (neonknight64) [inline applications/tari_base_node/src/main.rs]:
```suggestion
    handle_ctrl_c(&rt, flag);
```

File: base_layer/core/src/mining/miner.rs [hunk 2]
```diff
@@ -34,18 +35,31 @@ use crate::{
...
     time,
 };
+use tari_broadcast_channel::Subscriber;
 use tari_crypto::keys::SecretKey;
+use tokio::runtime;
 
 pub struct Miner<B: BlockchainBackend> {
     kill_flag: Arc<AtomicBool>,
-    stop_mining_flag: Arc<AtomicBool>,
+    got_new_block_event_flag: Arc<AtomicBool>,
```
Comment (neonknight64) [inline base_layer/core/src/mining/miner.rs]: Maybe `received_new_block_event_flag` is a better name.

Comment (CjS77) [reply base_layer/core/src/mining/miner.rs]: or `received_new_block_flag`

File: applications/tari_base_node/src/main.rs (hunk 1 above)
Comment (SWvheerden) [reply applications/tari_base_node/src/main.rs]: done

File: base_layer/core/src/mining/miner.rs (hunk 2 above)
Comment (SWvheerden) [reply base_layer/core/src/mining/miner.rs]: done

File: applications/tari_base_node/src/main.rs [hunk 3]
```diff
@@ -133,6 +133,12 @@ fn main() {
     let flag = node.get_flag();
     handle_ctrl_c(&rt, flag);
 
+    // lets run the miner
+    // Todo, enable this on config
+    if node_config.enable_base_node_mining {
```
[the remark this answers is not in the document]
Comment (SWvheerden) [reply applications/tari_base_node/src/main.rs]: done

File: base_layer/core/src/mining/miner.rs [hunk 4]
```diff
@@ -54,47 +68,83 @@ impl<B: BlockchainBackend> Miner<B> {
...
+                        BlockEvent::Verified((_, result)) => {
+                            if result == BlockAddResult::Ok{
+                                flag.store(true, Ordering::Relaxed);
+                            }
+                        },
+                        _ => (),
+                    }
+                },
+                (_) = t_miner => {break;
+                                },
```
[the remark this answers is not in the document]
Comment (SWvheerden) [reply base_layer/core/src/mining/miner.rs]: done

File: common/src/configuration.rs [hunk 5]
```diff
@@ -123,6 +123,7 @@ pub struct GlobalConfig {
     pub address: String,
     pub peer_seeds: Vec<String>,
     pub peer_db_path: String,
+    pub enable_base_node_mining: bool,
```
[the remark this answers is not in the document]
Comment (SWvheerden) [reply common/src/configuration.rs]: done

File: common/src/configuration.rs [hunk 6]
```diff
@@ -260,6 +269,8 @@ pub fn default_config() -> Config {
     cfg.set_default("base_node.mainnet.grpc_enabled", false).unwrap();
     cfg.set_default("base_node.mainnet.grpc_address", "tcp://127.0.0.1:18041")
         .unwrap();
+    cfg.set_default("base_node.mainnet.enable_base_node_mining", false)
```
[the remark this answers is not in the document]
Comment (SWvheerden) [reply common/src/configuration.rs]: done

File: common/src/configuration.rs [hunk 7]
```diff
@@ -279,6 +290,8 @@ pub fn default_config() -> Config {
     cfg.set_default("base_node.testnet.grpc_enabled", false).unwrap();
     cfg.set_default("base_node.testnet.grpc_address", "tcp://127.0.0.1:18141")
         .unwrap();
+    cfg.set_default("base_node.testnet.enable_base_node_mining", true)
```
[the remark this answers is not in the document]
Comment (SWvheerden) [reply common/src/configuration.rs]: done

**Pull request merged**